File: stats.js
var Q = require('q');
var discourse = require('./discourse.js').discourse;

function get_topic_posts(topic_id){
    return Q.nfcall(discourse.getTopic, topic_id).then(function(topic){
        var promises = [];
        topic.post_stream.stream.forEach(function(post_id, index){
            promises.push(Q.nfcall(discourse.getPost, post_id));
            console.log('Fetching post ' + post_id + ' (' + (1+index) + '/' + topic.post_stream.stream.length + ')');
        });
        return Q.all(promises);
    });
}

Q.nfcall(discourse.getLatest).then(function(topics){
    console.log(topics.length);
    var promises = [];
    topics.forEach(function(topic, index){
        console.log('Fetching topic ' + topic.id + ' (' + (1 + index) + '/' + topics.length + ')');
        promises.push(get_topic_posts(topic.id));
    });
    return Q.all(promises);
}).then(function(posts){
    console.log('Finally, all posts');
    console.log(posts);
}).catch(function (error) {
        // Handle any error from all above steps
    console.log('error');
}).done();

File: stats.py
import discourse
from config import config

def getDiscourseStats(forum_url):
    forum_config = config.forums.get(forum_url)
    credentials = {
        'api_key': forum_config.get('api_key'),
        'api_username': forum_config.get('api_username')
    }
    api = discourse.DiscourseApi(forum_url, credentials)
    posts = []
    post_ids = []
    topics = api.getAllTopics()
    for topic in topics:
        print('Get topic {0} data'.format(topic['id']))
        topic_data = api.getTopic(topic['id'])
        post_ids += topic_data['post_stream']['stream']
    for post_id in post_ids:
        print('Get post {0} data'.format(post_id))
        post_data = api.getPost(post_id)
        posts += [post_data]
    return posts, topics
